Allow generating migrations from within umbrella app

mpotra opened the pull request:
While numerous ecto_sql tasks are allowed to run from the root of an umbrella app when given `--migrations-path`, the `ecto.gen.migration` task is still prevented to run due to a `no_umbrella!/1` check.

This PR moves the umbrella check under a missing `--migrations-path` check.

If the mix task is run without a `--migrations-path` option, then the `gen.migration` task will fail as always.
However, with a `--migrations-path` option set, the task is allowed to run

wojtekmach commented:
I think the reason we don't allow running this task from the umbrella root is you can have multiple apps, each with their own repo and migrations folder, and then it would ambitious where to put this new migration.

mpotra commented:
> I think the reason we don't allow running this task from the umbrella root is you can have multiple apps, each with their own repo and migrations folder, and then it would be ambiguous where to put this new migration.

It makes perfect sense, but using `--migrations-path` and `--repo` command args would still be valid and should not be prevented from running IMHO; that is why this PR only allows running the command with `--migrations-path`, otherwise it's still prevented

josevalim commented:
Is it possible to add a test? 🤔

wojtekmach commented:
I take this back, when running this from umbrella root, there's no ambiguity where to generate the migration, we'd simply generate the migration for all apps. Strictly speaking we generate migrations for all _repos_, they could come from different apps or the same app could have multiple.

In that case I don't think we need to fail when the migrations path is not given, I think it's ok to use the default migration path for that repo. I think this PR could simply remove the `no_umbrella!("ecto.gen.migration")` check (+ add a test if possible). WDYT?

wojtekmach commented:
In fact when we have multiple repos we'd fail _if_ migrations path is given because it'd try to write the same file twice. :)

mpotra commented:
Thank you both for your feedback!

I've removed the `--migrations-path` requirement and the `no_umbrella!/1` check from `gen.migration`, as per @wojtekmach comments above.

Added a test to verify that it does not raise any longer under umbrella app.
 _Is my approach on emulating the umbrella app OK for the test?_

josevalim commented on lib/mix/tasks/ecto.gen.migration.ex:
```diff
@@ -61,7 +61,6 @@ defmodule Mix.Tasks.Ecto.Gen.Migration do
 
   @impl true
   def run(args) do
-    no_umbrella!("ecto.gen.migration")
```
We still want to raise if --migrations-path is given and there are multiple repos, no?

josevalim replied:
Ignore me!

josevalim commented on test/mix/tasks/ecto.gen.migration_test.exs:
```diff
@@ -68,4 +68,27 @@ defmodule Mix.Tasks.Ecto.Gen.MigrationTest do
     assert_raise Mix.Error, fn -> run ["-r", to_string(Repo)] end
   end
+
+  test "generates migration within umbrella app without raising" do
```
The test is good but I am afraid it is too brittle. So I think we should rather remove it... sorry! :D

josevalim commented:
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
